cirq/value/linear_dict.py

```diff
@@ -0,0 +1,265 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Linear combination represented as mapping of things to coefficients."""
+
+from typing import (Any, Callable, Dict, ItemsView, Iterable, Iterator,
+                    KeysView, Mapping, overload, Tuple, TypeVar, Union,
+                    ValuesView)
+
+Scalar = Union[complex, float]
+TVector = TypeVar('TVector')
+
+_TDefault = TypeVar('_TDefault')
+
+
+class LinearDict(Dict[TVector, Scalar]):
+    """Represents linear combination of things.
+
+    LinearDict implements the basic linear algebraic operations of vector
+    addition and scalar multiplication for linear combinations of abstract
+    vectors. Keys represent the vectors, values represent their coefficients.
+    The only requirement on the keys is that they be hashable (i.e. are
+    immutable and implement __hash__ and __eq__ with equal objects hashing
+    to equal values).
+
+    A consequence of treating keys as opaque is that all relationships between
+    the keys other than equality are ignored. In particular, keys are allowed
+    to be linearly dependent.
+    """
+    def __init__(self, terms: Mapping[TVector, Scalar]) -> None:
+        """Initializes linear combination from a collection of terms.
+
+        Args:
+            terms: Mapping of abstract vectors to coefficients in the linear
+                combination being initialized.
+        """
+        super().__init__()
+        self.update(terms)
+
+    @classmethod
+    def fromkeys(cls, vectors, coefficient=0):
+        return LinearDict(dict.fromkeys(vectors, complex(coefficient)))
+
+    def clean(self, *, atol: float=1e-9) -> 'LinearDict':
+        negligible = [v for v, c in super().items() if abs(c) <= atol]
+        for v in negligible:
+            del self[v]
+        return self
+
+    def copy(self) -> 'LinearDict':
+        return LinearDict(super().copy())
+
+    def keys(self) -> KeysView[TVector]:
+        snapshot = self.copy().clean(atol=0)
+        return super(LinearDict, snapshot).keys()
+
+    def values(self) -> ValuesView[Scalar]:
+        snapshot = self.copy().clean(atol=0)
+        return super(LinearDict, snapshot).values()
+
+    def items(self) -> ItemsView[TVector, Scalar]:
+        snapshot = self.copy().clean(atol=0)
+        return super(LinearDict, snapshot).items()
```

Comment:
`super()`

Reply:
This would pick up `self` instead of `snapshot`.

(This would ultimately lead to `RuntimeError` being raised when one iterates over a LinearDict and happens to set a coefficient to zero (since this removes the element from the underlying dictionary). Confirmed by making the change and seeing the corresponding unit test fail.)